spec.bs

```diff
@@ -2354,7 +2340,12 @@ of the following global objects:
       given |perBuyerSignals| if |perBuyerSignals| is not null, otherwise {{undefined}}.
     1. Let |trustedBiddingSignalsJS| be |trustedBiddingSignals| [=converted to ECMAScript values=].
     1. Let |browserSignalsJS| be |browserSignals| [=converted to ECMAScript values=].
-    1. Let |directFromSellerSignalsJs| be |buyerDirectFromSellerSignals| [=converted to ECMAScript values=].
+    1. Let |directFromSellerSignalsJs| be {{undefined}}.
+    1. If |directFromSellerSignalsForBuyer| is not null:
+      1. Let |directFromSellerSignalsString| be the result of
+        [=serializing an Infra value to a JSON string=] given |directFromSellerSignalsForBuyer|.

```

Comment:
I think this isn't quite right?

So, the directFromSellerSignals object is always passed to worklets, even if no signals were provided. It's just that its members (perBuyerSignals / auctionSignals or sellerSignals / auctionSignals) are set to null if not provided -- they're never undefined. For each field, if provided, each of perBuyerSignals / auctionSignals or sellerSignals / auctionSignals is the result of turning the signals for that field into a JavaScript value (object, array, number, string, etc.)

See the implementation for buyers [0] and for sellers [1] (the reporting functions, in those same files, are also very similar).

[0] https://source.chromium.org/chromium/chromium/src/+/main:content/services/auction_worklet/bidder_worklet.cc;l=1371;drc=f3ee53147a677fb8d5fac5a83e68a41041c66758

[1] https://source.chromium.org/chromium/chromium/src/+/main:content/services/auction_worklet/seller_worklet.cc;l=1014;drc=f3ee53147a677fb8d5fac5a83e68a41041c66758

Reply:
Right, the {{undefined}} is not correct.
What's the [v8::Null()](https://crsrc.org/c/content/services/auction_worklet/bidder_worklet.cc;drc=150d8c7e45daeef094be8ec8852e3486eed8f59d;l=107) converted to when passed to worklets, when there's no direct from seller signals? This v8::Null seems to be the whole directFromSellerSignals, not for a single field of it.

Reply:
So, in the implementation code I linked, `GetDirectFromSellerSignals()` can return `v8::Null`, but that only applies to one of `per_buyer_signals` / `auction_signals` (for bidder_worklet.cc) or `seller_signals` / `auction_signals` (for seller_worklet.cc), not the whole directFromSellerSignals object passed as a function parameter.

Note that `direct_from_seller_signals` / `direct_from_seller_signals_dict` is always created (that's the directFromSellerSignals function parameter, although JavaScript doesn't have named function arguments), and always pushed onto `args`, and it always has members `perBuyerSignals` / `auctionSignals` (bidder_worklet.cc) or `sellerSignals` / `auctionSignals` (seller_worklet.cc).

Reply:
Oh my bad. I didn't read the caller that you linked, but just the [GetDirectFromSellerSignals](https://source.chromium.org/chromium/chromium/src/+/main:content/services/auction_worklet/bidder_worklet.cc;drc=150d8c7e45daeef094be8ec8852e3486eed8f59d;l=108) method, and simply thought it was for the whole directFromSellerSignals object based on its function name.
Now I see how it works. Thanks!